### PR TITLE
moving remeshing to optionally use rabbitmq celery workers and message broadcasting

### PR DESCRIPTION
**Pull request by fcollman**
This will decouple the remeshing workers from the pcg, allowing them to scale differently if needed, and also adds broadcasting of the remeshing events to a rabbitmq exchange.  compartment classification workers might declare a queue on that exchange to know when they might need to rerun themselves.   the remeshing itself now takes an optional argument to use a celery worker (Default no for now).  We can test that this works before merging but i wanted to get your eyes on it early.

### CODE REVIEW

File: pychunkedgraph/backend/chunkedgraph.py [hunk 1]
```diff
@@ -3054,8 +3055,14 @@ def add_edges(self, user_id: str, atomic_edges: Sequence[np.uint64],
...
-                        meshgen.mesh_lvl2_previews(self, list(
-                            lvl2_node_mapping.keys()))
+                        if use_mesh_celery_worker:
+                            serialized_cg_info = self.get_serialized_info()
+                            del serialized_cg_info["credentials"]
+
+                            worker.mesh_lvl2_preview(serialized_cg_info, )
+                        else:
+                            meshgen.mesh_lvl2_previews(self, list(
+                                lvl2_node_mapping.keys()))
```
Comment (sdorkenw) [inline pychunkedgraph/backend/chunkedgraph.py]: There is an additional location where preview remeshing is triggered down in remove_edges.

File: pychunkedgraph/meshing/meshgen.py [hunk 2]
```diff
@@ -367,20 +368,24 @@ def mesh_lvl2_previews(cg, lvl2_node_ids, cv_path=None,
...
-        multi_args.append([serialized_cg_info, lvl2_node_id,
-                           lvl2_node_ids[lvl2_node_id],
-                           cv_path, cv_mesh_dir, mip, simplification_factor,
-                           max_err, parallel_download, verbose,
-                           cache_control])
+            multi_args.append([serialized_cg_info, lvl2_node_id,
+                            lvl2_node_ids[lvl2_node_id],
+                            cv_path, cv_mesh_dir, mip, simplification_factor,
+                            max_err, parallel_download, verbose,
+                            cache_control])
```
Comment (sdorkenw) [inline pychunkedgraph/meshing/meshgen.py]: here some formatting seems to be off

File: pychunkedgraph/meshing/meshgen.py [hunk 3]
```diff
@@ -367,20 +368,24 @@ def mesh_lvl2_previews(cg, lvl2_node_ids, cv_path=None,
...
     if n_threads == 1:
         mu.multiprocess_func(_mesh_lvl2_previews_threads,
                              multi_args, n_threads=n_threads,
                              verbose=False, debug=n_threads==1)
     else:
-        mu.multisubprocess_func(_mesh_lvl2_previews_threads,
-                                multi_args, n_threads=n_threads)
+        if use_celery_worker:
+            res = mesh_lvl2_previews_task.chunks(multi_args, n_threads)
+            res.get()
```
Comment (sdorkenw) [inline pychunkedgraph/meshing/meshgen.py]: this should be done asynchronously

File: pychunkedgraph/meshing/worker.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#!/usr/bin/env python3
+
+from chunkedgraph.meshing.meshgen import mesh_lvl2_preview
```
Comment (nkemnitz) [inline pychunkedgraph/meshing/worker.py]:
```suggestion
from pychunkedgraph.meshing.meshgen import mesh_lvl2_preview
```

File: pychunkedgraph/meshing/worker.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import chunkedgraph
```
Comment (nkemnitz) [inline pychunkedgraph/meshing/worker.py]:
```suggestion
import pychunkedgraph
```

File: pychunkedgraph/meshing/worker.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+# for our default
+
+broker_url = 'amqp://%s:%s@%s/%s' % (rabbitmq_user,
+                                     rabbitmq_password,
+                                     broker_service_host,
+                                     rabbitmq_vhost)
+# this sets this module up as a celery worker app
+app = Celery('tasks', broker=broker_url, backend='amqp')
+
+# this sets up the pike client to do a broadcast of the remeshing event
```
Comment (nkemnitz) [inline pychunkedgraph/meshing/worker.py]:
```suggestion
# this sets up the pika client to do a broadcast of the remeshing event
```

File: requirements.txt [hunk 7]
```diff
@@ -11,3 +11,4 @@ multiwrapper
 cloud-volume
 python-json-logger
 igneous==dev
+celery
```
Comment (nkemnitz) [inline requirements.txt]: probably missing `pika` here.

Comment (fcollman) [reply requirements.txt]: fixed

File: pychunkedgraph/meshing/worker.py (hunk 6 above)
Comment (fcollman) [reply pychunkedgraph/meshing/worker.py]: fixed

File: pychunkedgraph/meshing/worker.py (hunk 5 above)
Comment (fcollman) [reply pychunkedgraph/meshing/worker.py]: fixed and refactored

File: pychunkedgraph/meshing/worker.py (hunk 4 above)
Comment (fcollman) [reply pychunkedgraph/meshing/worker.py]: fixed and refactored

**Comment by fcollman** [comment]
so in this case, you can have as many celery workers as you want servicing these celery job calls.  So right now in practice we would start another copy of the same docker image as a separate pod, but it  would run a celery worker command instead.  Then we could control the number of replicas in that pod independent of the number of replicas in the pychunkedgraph.  I like the idea of separating the pcg from the meshing further.  Maybe in fact the pcg should just publish and event that it has updated chunk X, and other services can subscribe that event and do the appropriate thing.  The only tricky part is that for non preview remeshing one needs to know what the adjacent chunks are that were affected by the preview meshing so that remeshing can pull them all in.. since the pcg is the source of truth about what chunks are adjacent i'm not sure how to pull that off.  Note that skeletons could work in a similar way.

**Comment by nkemnitz** [comment]
A Chunk Update event would be great. Doing the meshing with that should be fine: The Chunk IDs are chosen so that you can easily calculate the adjacent chunk IDs, so no problem there.

The two things the mesher really needs from such a chunk update event are:
1) the low-level Chunk IDs where edges have been added/removed.
2) the root ID(s) of the newly generated object(s) - only want to do sparse meshing.
Everything else, such as neighboring chunks, the mesher can retrieve on its own using the PyChunkedGraph module. I just don't want the PyChunkedGraph to contain any meshing dependencies. The other way around (Mesher using PyChunkedGraph to fetch information) is fine, I think.

**Comment by nkemnitz** [comment]
@fcollman Was that PR intented to merged ASAP, or would you like to continue working on this branch?

**Comment by fcollman** [comment]
my intent was to start a discussion, I don't think this is ready for merging.  Needed you and sven to be on board with the direction.

**Pull request closed without merging**